app/controllers/result_filters/location_controller.rb

```diff
@@ -2,21 +2,89 @@ module ResultFilters
   class LocationController < ApplicationController
     include FilterParameters
 
+    NO_OPTION = nil
+    LOCATION_OPTION = "1".freeze
+
     def new; end
 
     def create
-      if(!params[:l])
-        flash[:error] = "Please choose an option"
-        redirect_to location_path(params_without_unsupported_location_params)
+      build_params = strip(filter_params.clone)
+
+      case selected_option
+      when NO_OPTION
+        handle_no_option(build_params)
+      when LOCATION_OPTION
+        if location_query.empty?
+          handle_missing_location(build_params)
+        else
+          handle_location_option(build_params)
+        end
       else
-        redirect_to results_path(params_without_unsupported_location_params)
+        redirect_to results_path(build_params)
       end
     end
 
   private
 
-    def params_without_unsupported_location_params
-      filter_params.except(:lat, :lng, :rad, :query, :loc, :lq)
+    def handle_location_option(build_params)
+      geocode_params = geocode_params_for(location_query)
+      if geocode_params
+        merged_params = build_params.merge(geocode_params)
+        redirect_to results_path(merged_params)
+      else
+        build_errors_for(:unknown_location)
+        redirect_to location_path(build_params)
+      end
+    end
+
+    def handle_missing_location(build_params)
+      build_errors_for(:missing_location)
+      redirect_to location_path(build_params)
+    end
+
+    def handle_no_option(build_params)
+      build_errors_for(:none_selected)
+      redirect_to location_path(build_params)
+    end
+
+    def build_errors_for(type)
+      flash[:query] = location_query
+
+      case type
+      when :none_selected
+        flash[:error] = "Please choose an option"
+      when :missing_location
+        flash[:error] = "Postcode, town or city"
+        flash[:location_error] = "Please enter a postcode, city or town in England"
```

Comment:
```suggestion
        flash[:location_error] = "Please enter a postcode, town or city in England"
```

![image](https://user-images.githubusercontent.com/470137/73652013-af0de080-467d-11ea-8588-fb34baf5bd61.png)

I think you ported c# bug along...